compiler/rustc_windows_rc/src/lib.rs

```diff
@@ -0,0 +1,145 @@
+//! A build script dependency to create a Windows resource file for the compiler
+//!
+//! Uses values from the `CFG_VERSION` and `CFG_RELEASE` environment variables
+//! to set the product and file version information in the Windows resource file.
+use std::{env, ffi, fs, path, process};
+
+use cc::windows_registry;
+
+/// The template for the Windows resource file.
+const RESOURCE_TEMPLATE: &str = include_str!("../rustc.rc.in");
+
+const VFT_APP: u32 = 0x00000001; // VFT_APP
+const VFT_DLL: u32 = 0x00000002; // VFT_DLL
+
+#[derive(Debug, Clone, Copy)]
+#[repr(u32)]
+pub enum VersionInfoFileType {
+    App = VFT_APP,
+    Dll = VFT_DLL,
+}
+
+/// Create and compile a Windows resource file with the product and file version information for the rust compiler.
+///
+/// Returns the path to the compiled resource file
+///
+/// Does not emit any cargo directives, the caller is responsible for that.
+pub fn compile_windows_resource_file(
+    file_stem: &path::Path,
+    file_description: &str,
+    filetype: VersionInfoFileType,
+) -> path::PathBuf {
+    let mut resources_dir = path::PathBuf::from(env::var_os("OUT_DIR").unwrap());
+    resources_dir.push("resources");
+    fs::create_dir_all(&resources_dir).unwrap();
+
+    let resource_compiler = find_resource_compiler(&env::var("CARGO_CFG_TARGET_ARCH").unwrap())
+        .expect("rc.exe not found");
+
+    let mut resource_script = RESOURCE_TEMPLATE.to_string();
+
+    // Set the string product version to the same thing as `rustc --version`
+    let product_version = env::var("CFG_VERSION").unwrap_or("unknown".to_string());
+    let file_version = &product_version;
+
+    // This is just "major.minor.patch" and a "-dev", "-nightly" or similar suffix
+    let rel_version = env::var("CFG_RELEASE").unwrap();
+
+    let product_name = product_name(env::var("CFG_RELEASE_CHANNEL").unwrap());
```

Comment:
`product_name` is between code for handling CFG_RELEASE, so perhaps reorder that?